### PR TITLE
Remove all AMP logic from bundle

### LINKED ISSUE #2233: Remove all AMP logic from bundle
**Is your feature request related to a problem? Please describe.**
All our AMP logic is included in the bundle and therefore clientside render. This is completely wasteful in bundle size and JS execution time.

**Describe the solution you'd like**
Cut the AMP logic from the bundle on build

**Describe alternatives you've considered**
Perhaps a similar approach to the one taken by this library could be used: transform-react-remove-prop-types

**Testing notes**
Regression testing of amp pages.

Dev insight: no unit test changes required

**Additional context**
n/a

### PR DESCRIPTION
**Pull request by FK78**
Resolves #2233 

**Overall change:** _Cut the AMP logic from the bundle on build._

**Code changes:**

- _Use [Loadable Components](https://github.com/smooth-code/loadable-components) to remove AMP logic from the bundle on build._

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
- [ ] This PR requires manual testing

### CODE REVIEW

**Comment by jroebu14** [comment]
If the `IgnorePlugin` doesn't add to the bundle any import with `amp` in the path, then would this not prevent our AMP service from working as expected? Also, if anything with `amp` is not loaded then surely there's an error when using the AMP components in components? I've based my assumptions of this https://remarkablemark.org/blog/2017/02/25/webpack-ignore-module/

**Comment by dr3** [comment]
Im a tad concerned with the ignore plugin, if we accidently use amp on the client it might blow the SPA up and throw an error, im not sure. We likely need a safer way to do it

**Comment by dr3** [comment]
My suggestion is to create files that we would load on the client bundles instead.

E.g. if we had `Amp/index.jsx` in a component, we would create `Amp/index.empty.jsx` or `Amp/index.client.js`

These files would sit next to files we dont want in the client bundles. The new files would mirror the exports of the main file. E.g. replace react component with `() => null`

When webpack client runs, we can check each file to see if it has a sister file (the empty/client one) and use that instead

```
new webpack.NormalModuleReplacementPlugin(
  /.*/,
  resource => {
      const [prefix, suffix] = resource.request.split('.js');
      const empty = `${prefix}.empty.js${suffix}`;

      if (fs.existsSync(empty)) {
        resource.request = empty;
      }
  },
),
```

Pro:
Doesnt create bundles we will never use
Doesnt need explict e2es
Doesnt create risk of trying to use code that doesnt exist

Cons:
Adds extra files to our codebase

**Comment by FK78** [comment]
TL;DR

### Option 1: Use [Loadable Components](https://github.com/smooth-code/loadable-components).

**Pro**
>with the loadable components method it's clear that code splitting is happening (and clear that it should) and if a developer comes across it then they might start thinking about code splitting in other situations

>We are already testing that risk though right? We have to tests to ensure only one bundle is loaded and that it is the correct service bundle.
We can extend that to ensure an amp bundle is not loaded.

>On the other hand it’s probably the least complex solution.

>This is essentially code splitting based on the route. It just so happens this route isn’t used client side.

>This is what code splitting is for.

**Con**
>We’d not be using the loadable as it is intended as we would be creating bundles we never intend to use.

>Creating client side bundles, that should never be loaded on the client side feels like much more of a hack, and a more risky one IMO

**Advice**
>I personally would say if we used loadable components then it should only be used on the psammead import of AMP components and not on the if() statements in Simorgh, this gives the majority of the AMP splitting benefit for the simplest solution

### Option 2: Use [Conditional Imports](https://levelup.gitconnected.com/decreasing-webpack-bundle-size-with-conditional-imports-f8bb94db8195), [another link](https://bbcnews.slack.com/archives/GHNLB0MMK/p1572533435072100?thread_ts=1572517488.055500&cid=GHNLB0MMK).

**Pro**
>Means that we dont get the weight of the amp code, we wont break as the file API is consistant and we dont risk having bundles we accidently load on the client and dont know about

>`.empty.` is a bit hacky but does already have precedence in simorgh with the `logger.node.js`, however we would need to clearly explain this in an ADR and make it a clear part of the AMP component workflow. I think this use of Webpack replace plugin is probably the most complete solution but also the most complex.

>I agree that .empty. feels like a hack, but its intentionally removing code from our bundles.

**Con**
>The .empty file method means extra work than can easily be missed every time we add AMP functionality

>Adds extra files to our codebase

>feels like a hack.

Some more [opinions](https://bbcnews.slack.com/archives/GHNLB0MMK/p1572517488055500).

**Comment by ryanmccombe** [comment]
Seems this problem could be solved once rather than every time we want to use amp-specific code?

eg, ignoring imports ending in `.amp.jsx` in the bundle that is sent to clients, but keeping it on the server?

Edit: Fahad wanted me to elaborate more on this option:

First, we can remove or split out any file that ends in 'amp.js' using the standard webpack config API - either `splitChunks` or `ignorePlugin`

Second, we can then include those files exclusively when rendering on the server.

I'm not too clear what would be the easiest way to do that second part - we use `loadable/server` to determine what chunks to use for SSR.  I don't have much experience with that library, but if it's not straightforward to do it through that, it would seem to be a solved problem more generally.

It does seem like "running code only for SSR" will be an ongoing requirement, not just for `amp`, so it seems like it's a problem worth solving once, in a standard way

**Comment by jamesdonoh** [comment]
Has anyone quantified the actual potential savings of doing this? The issue says _All our AMP logic is included in the bundle_ - this is not ideal but does the expected bundle size decrease justify the added complexity of the solutions described above?

**Comment by ryanmccombe** [comment]
> Has anyone quantified the actual potential savings of doing this?

It's just this file: https://github.com/bbc/simorgh/blob/9f80f52ed8a683b931dbe5485db673b41816c0af/src/app/containers/ConsentBanner/index.amp.jsx - about 1kb before gzip

**Comment by FK78** [comment]
Closing PR as the added complexity of the solutions do not justify the actual savings.
